services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/AnnotationsTranslatorImpl.java

```diff
@@ -0,0 +1,379 @@
+	 * 
+	 * @param index
+	 * @param array
+	 * @return
+	 */
+	Optional<ListValue> attemptToReadAsLong(int index, JSONArray array) {
+		try {
+			Long value = array.getLong(index);
+			return Optional.of(new ListValue(AnnotationsValueType.LONG, value.toString()));
+		} catch (JSONException e) {
+			return Optional.empty();
+		}
+	}
+
+	/**
+	 * Attempt to read the value for the given key as a JSONArray. If the value is
+	 * not a JSONArray an empty optional will be returned.
+	 * 
+	 * @param key
+	 * @param jsonObject
+	 * @return
+	 */
+	Optional<AnnotationsValue> attemptToReadAsJSONArray(String key, JSONObject jsonObject) {
+		try {
+			JSONArray array = jsonObject.getJSONArray(key);
+			List<String> valueList = new ArrayList<String>(array.length());
+			AnnotationsValueType lastType = null;
+			for (int i = 0; i < array.length(); i++) {
+				ListValue listValue = Stream
+						.of(attemptToReadAsDouble(i, array), attemptToReadAsTimestamp(i, array),
+								attemptToReadAsLong(i, array), attemptToReadAsString(i, array))
+						.filter(Optional::isPresent).findFirst().get().get();
+				if (lastType != null && !lastType.equals(listValue.getType())) {
+					throw new IllegalArgumentException("List of mixed types found for key: '" + key + "'");
+				}
+				lastType = listValue.getType();
+				valueList.add(listValue.getValue());
+			}
+			AnnotationsValue annValue = new AnnotationsValue();
+			annValue.setType(lastType);
+			annValue.setValue(valueList);
+			return Optional.of(annValue);
+		} catch (JSONException e) {
+			return Optional.empty();
+		}
+	}
+
+	/**
+	 * Field names of an Entity cannot be used as annotation keys. All other keys
+	 * are valid.
+	 * 
+	 * @param clazz
+	 * @param key
+	 * @return
+	 */
+	public static boolean canUseKey(Class clazz, String key) {
+		try {
+			clazz.getDeclaredField(key);
+			return false;
+		} catch (NoSuchFieldException e) {
+			return true;
+		} catch (SecurityException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	/**
+	 * Write the provided annotations to the provided JSONObject. Note: The
+	 * resulting JSON format is not the same as the format used to write Annotations
+	 * in API requests.
+	 * 
+	 * @param toWrite
+	 * @param jsonObject
+	 */
+	void writeAnnotationsToJSONObject(Annotations toWrite, JSONObject jsonObject) {
+		for (Entry<String, AnnotationsValue> entry : toWrite.getAnnotations().entrySet()) {
+			writeAnnotationValue(entry.getKey(), entry.getValue(), jsonObject);
+		}
+	}
+
+	void writeAnnotationValue(String key, AnnotationsValue value, JSONObject jsonObject) {
+		if (value.getValue() != null) {
+			if (value.getValue().size() > 1) {
+				writeAnnotationValueList(key, value.getType(), value.getValue(), jsonObject);
+			} else {
+				writeAnnotationSingle(key, value.getType(), value.getValue().get(0), jsonObject);
+			}
+		}
+	}
+
+	void writeAnnotationSingle(String key, AnnotationsValueType type, String value, JSONObject jsonObject) {
+		switch (type) {
```

Comment:
We can extract the logic to translate the "String value" to a JSON String value by type, having a single toJsonValue(AnnotationValueType type, String value) that can be reused for both the single and multiple case.